core/rawdb/freezer_table.go

```diff
@@ -546,84 +559,179 @@ func (t *freezerTable) append(item uint64, encodedBlob []byte, wlock bool) (bool
 	return false, nil
 }
 
-// getBounds returns the indexes for the item
-// returns start, end, filenumber and error
-func (t *freezerTable) getBounds(item uint64) (uint32, uint32, uint32, error) {
-	buffer := make([]byte, indexEntrySize)
-	var startIdx, endIdx indexEntry
-	// Read second index
-	if _, err := t.index.ReadAt(buffer, int64((item+1)*indexEntrySize)); err != nil {
-		return 0, 0, 0, err
+// getIndices returns the index entries for the given from-item, covering 'count' items.
+// N.B: The actual number of returned indices for N items will always be N+1 (unless an
+// error is returned).
+func (t *freezerTable) getIndices(from, count uint64) ([]*indexEntry, error) {
+	// Apply the table-offset
+	from = from - uint64(t.itemOffset)
+	// For reading N items, we need N+1 indices.
+	buffer := make([]byte, (count+1)*indexEntrySize)
+	if _, err := t.index.ReadAt(buffer, int64(from*indexEntrySize)); err != nil {
+		return nil, err
 	}
-	endIdx.unmarshalBinary(buffer)
-	// Read first index (unless it's the very first item)
-	if item != 0 {
-		if _, err := t.index.ReadAt(buffer, int64(item*indexEntrySize)); err != nil {
-			return 0, 0, 0, err
-		}
-		startIdx.unmarshalBinary(buffer)
-	} else {
+	var (
+		indices []*indexEntry
+		offset  int
+	)
+	for i := from; i <= from+count; i++ {
+		var startIndex = new(indexEntry)
+		startIndex.unmarshalBinary(buffer[offset:])
+		offset += indexEntrySize
+		indices = append(indices, startIndex)
+	}
+	if from == 0 {
 		// Special case if we're reading the first item in the freezer. We assume that
 		// the first item always start from zero(regarding the deletion, we
 		// only support deletion by files, so that the assumption is held).
 		// This means we can use the first item metadata to carry information about
 		// the 'global' offset, for the deletion-case
-		return 0, endIdx.offset, endIdx.filenum, nil
+		indices[0].offset = 0
+		indices[0].filenum = indices[1].filenum
```

Comment:
I'm not really following. Isn't the offset already zero if it's the first item? I also don't understand why the file number is assigned?

Reply:
Alternatively, isn't this the case for all items that start a new file, not just `from == 0`?

Reply:
Or, is the point that "indices[0].filenum` might not exist at all (since it's a deleted file), vs. anything else exists and we can "read and see the data overflown"?

Reply:
It's more or less what the comment says above. So each index is a pair of `filenum, offset`. Like this: 
```
| number | fileno | offset |
|--------|--------|--------|
|  000   |  000   |  000   | 
|  001   |  000   |  020   | 
|  002   |  000   |  040   | 
|  003   |  001   |  020   | 
|  004   |  001   |  040   | 
|  005   |  002   |  020   |  <-- we want to delete anything before here
|  006   |  002   |  040   | 
```
However, the `0`th index will _always_ be `|  000   |  000   |  000   |`, right? 

Therefore, back in the day, we decided to use this quirk to implement deletability. So if we want to delete the file `0` and `1`, we need to store the _item_-offset somewhere. So we store that in the `offset` (which otherwise is the _byte_-offset within the file -- but we know the first item in a file always has byte offset `0`). 
```
| number | fileno | offset |
|--------|--------|--------|
|  000   |  002   |  004   | 
|  001   |  002   |  020   | 
|  002   |  002   |  040   | 
```

I think you're right though, that we wouldn't really need to set the filenum. However, as you can see in the previous code, for item `N`, marked by `indexA` and `indexB`, the actual data is always stored in the file given by `indexB`. The two may differ, if the data crosses a file boundary. 

In this case, we want to return to the caller info about where to start reading, and where to stop. Therefore, doing this conversion here simplifies for the calling code, which can just treat the indices as direct data information without dealing with the intricacies of the data indexing scheme.